Wrap basic types in a struct

liamsi opened the pull request:
ref #204

Note: Instead of switching over all possible cases via reflection (and maybe recursively), this simply adds the bytes. This PR just emulates a wrapping struct.

liamsi committed:
First stab on #204: disallow direct encoding of basic types

liamsi committed:
fix some tests and disallow defined types on decoding
(but allow interfaces on decoding)

liamsi committed:
fix remaining tests and introduce a new error type (instead of panicing)

liamsi committed:
remove code: condition is always 'false' because 'err' is always 'nil'

liamsi committed:
WIP: add `(field_number << 3) | wire_type` with field_number == 1 if the
passed type is not a struct to simulate protobuf message {}

Still missing:

 - proper table driven tests
 - decoding (will basically be skipping over the first bytes if non struct type is passed)

liamsi committed:
remove unused type aliases (use types defined in tests package instead)

liamsi committed:
Some cleanup:
 - table driven tests
 - use helper method `writeFieldIfNotEmpty` where it makes sense
 - remove a bunch of comments & debug output

liamsi committed:
Added a more complex test case (currently fails), some lengthy debugging
session revealed another minor difference between amino and proto3 (see
the two added FIXMEs); also:
 - renamed some proto3 types
 - some minor cleanup

liamsi committed:
Do not prepend with list of structs with field tag (will be handled internally)

liamsi committed:
WIP: First stab on decoding -> a few failing tests

liamsi committed:
Fix one of the failing tests: TestBasicTypes

liamsi committed:
Fix all remaining tests

liamsi commented on amino.go:
```diff
@@ -203,21 +211,43 @@ func (cdc *Codec) MarshalBinaryBare(o interface{}) ([]byte, error) {
-		return nil, err
+	// in the case of of a repeated struct (e.g. type Alias []SomeStruct),
+	// we do not need to prepend with `(field_number << 3) | wire_type` as this
+	// would need to be done for each struct and not only for
```
```suggestion
	// would need to be done for each struct and not only for the first.
```

liamsi commented on amino.go:
```diff
@@ -387,10 +444,6 @@ func (cdc *Codec) MarshalJSON(o interface{}) ([]byte, error) {
-		// Part 2:
-		if err != nil {
-			return nil, err
-		}
```
Not sure what this was supposed to do ... The err would always have been nil here.

liamsi commented on json_test.go:
```diff
@@ -172,24 +172,8 @@ type innerFP struct {
 	cdc := amino.NewCodec()
-	// Binary doesn't support decoding to a map...
-	// TODO: move out binary tests from json_test.go ...
-	assert.Panics(t, func() {
```
These were not deleted but moved into binary_test.go

liamsi committed:
Fix unfinished comment & remove debug output

Signed-off-by: Ismail Khoffi <Ismail.Khoffi@gmail.com>

jordaaash commented on amino.go:
```diff
@@ -2,21 +2,29 @@ package amino
+	// NotPointerErr is thrown when you call a method that expects a pointer, e.g. Unmarshal
+	NotPointerErr = errors.New("expected a pointer")
+
+	UnknownBasicTypeErr = fmt.Errorf("got unknown type: ")
```
Is this error used anywhere?

<img width="571" alt="Screen Shot 2019-05-22 at 12 41 02 PM" src="https://user-images.githubusercontent.com/1173161/58195996-e6269800-7c8e-11e9-991b-0d4e97d59bbb.png">

liamsi replied:
Good catch! That's a relict from another approach (actually wrapping each basic type in a struct instead of just adding the necessary bytes). Removed.

jordaaash commented on amino.go:
```diff
@@ -2,21 +2,29 @@ package amino
-// we use this time to init. a zero value (opposed to reflect.Zero which gives time.Time{} / 01-01-01 00:00:00)
-var zeroTime time.Time
+	// NotPointerErr is thrown when you call a method that expects a pointer, e.g. Unmarshal
+	NotPointerErr = errors.New("expected a pointer")
```
I'm not familiar with stdlib's `errors` vs. `pkg/errors` vs. `fmt.Errorf`, but I saw this dependency isn't used elsewhere. Worth adding a new library for?

liamsi replied:
pkg/errors allows you to wrap the error with some context instead of just passing it on. This helps callers to understand exactly where the error occurred. Although you can achieve sth. similar with fmt.Errorf, I think error.Wrap is slightly more convenient (and it also adds the call stack if you want to).

liamsi replied:
This is an excellent read on errors and touches the different error packages too: https://dave.cheney.net/2016/04/27/dont-just-check-errors-handle-them-gracefully

liamsi committed:
remove unsed var

Signed-off-by: Ismail Khoffi <Ismail.Khoffi@gmail.com>

liamsi committed:
delete debug output and comment

Signed-off-by: Ismail Khoffi <Ismail.Khoffi@gmail.com>

liamsi commented on amino.go:
```diff
@@ -340,6 +371,28 @@ func (cdc *Codec) UnmarshalBinaryBare(bz []byte, ptr interface{}) error {
+			return errors.Wrap(err, "could not decode field number and type")
+		}
+		if fnum != 1 {
+			fmt.Println("What is it?", rv.Kind())
```
Forgot to delete this line.

liamsi commented on binary-encode.go:
```diff
@@ -413,27 +413,12 @@ func (cdc *Codec) encodeReflectBinaryStruct(w io.Writer, info *TypeInfo, rv refl
-					// not a pointer and empty:
-					buf.Truncate(lBeforeKey)
-				}
-
```
This logic was moved to `writeFieldIfNotEmpty` as it is now used in two places (when wrapping a basic type as a field of a struct).

liamsi committed:
delete debug output and obsolete TODOs and some minot cleanups

Signed-off-by: Ismail Khoffi <Ismail.Khoffi@gmail.com>

alexanderbez reviewed (approved):
ACK on first passthrough.

alexanderbez commented on amino.go:
```diff
@@ -340,6 +371,26 @@ func (cdc *Codec) UnmarshalBinaryBare(bz []byte, ptr interface{}) error {
 		}
 		bz = bz[4:]
 	}
+	isRepeatedStruct := info.Type.Kind() == reflect.Slice && info.Type.Elem().Kind() == reflect.Struct
```
Maybe we ought to have an auxiliary function given an `info` that returns a bool if it's a repeated struct.

liamsi replied:
Changed. Also, added another test case and a few related changes in 8764911

liamsi committed:
Add another test case (type PrimitivesStructAr [2]PrimitivesStruct)
and fix implementation:

 - Additionally to the fieldnum and wiretag we need a size prefix
   (not bare) in case the typ3 isn't of ByteLength.
 - add helper methods as Bez suggested: check if struct and check if
   pointer to struct
 - fix tests due to changes

Signed-off-by: Ismail Khoffi <Ismail.Khoffi@gmail.com>
